zebra/zebra_tc.c

```diff
@@ -0,0 +1,623 @@
+
+	key = jhash(&htb->rate, sizeof(htb->rate), 0);
+	key = jhash(&htb->ceil, sizeof(htb->ceil), key);
+
+	return key;
+}
+
+uint32_t zebra_tc_class_hash_key(const void *arg)
+{
+	const struct zebra_tc_class *class;
+	uint32_t key;
+
+	class = arg;
+
+	key = jhash_2words(class->class.ifindex, class->class.handle,
+			   class->class.kind);
+
+	switch (class->class.kind) {
+	case TC_QDISC_HTB:
+		key = tc_class_htb_hash_key(&class->class.u.htb);
+		break;
+	default:
+		break;
+	}
+
+	return key;
+}
+
+bool zebra_tc_class_hash_equal(const void *arg1, const void *arg2)
+{
+	const struct zebra_tc_class *c1, *c2;
+
+	c1 = (const struct zebra_tc_class *)arg1;
+	c2 = (const struct zebra_tc_class *)arg2;
+
+	if (c1->class.ifindex != c2->class.ifindex)
+		return false;
+
+	if (c1->class.handle != c2->class.handle)
+		return false;
+
+	if (c1->class.kind != c2->class.kind)
+		return false;
+
+	switch (c1->class.kind) {
+	case TC_QDISC_HTB:
+		if (c1->class.u.htb.rate != c2->class.u.htb.rate)
+			return false;
+		if (c1->class.u.htb.ceil != c2->class.u.htb.ceil)
+			return false;
+		break;
+	default:
+		break;
+	}
+
+	return true;
+}
+
+struct tc_class_handle_lookup {
+	struct zebra_tc_class *class;
+	uint32_t handle;
+};
+
+
+static int tc_class_lookup_handle_walker(struct hash_bucket *b, void *data)
+{
+	struct tc_class_handle_lookup *lookup = data;
+	struct zebra_tc_class *class = b->data;
+
+	if (lookup->handle == class->class.handle) {
+		lookup->class = class;
+		return HASHWALK_ABORT;
+	}
+
+	return HASHWALK_CONTINUE;
+}
+
+static struct zebra_tc_class *
+tc_class_lookup_handle(struct zebra_tc_class *class)
+{
+	struct tc_class_handle_lookup lookup;
+
+	lookup.handle = class->class.handle;
+	lookup.class = NULL;
+	hash_walk(zrouter.rules_hash, &tc_class_lookup_handle_walker, &lookup);
+
+	return lookup.class;
+}
+
+static void *tc_class_alloc_intern(void *arg)
+{
+	struct zebra_tc_class *class;
+	struct zebra_tc_class *new;
+
+	class = (struct zebra_tc_class *)arg;
+
+	new = XCALLOC(MTYPE_TMP, sizeof(*new));
+
+	memcpy(new, class, sizeof(*class));
+
+	return new;
+}
+
+static struct zebra_tc_class *tc_class_free(struct zebra_tc_class *hash_data,
+					    bool free_data)
+{
+	hash_release(zrouter.class_hash, hash_data);
+
+	if (free_data) {
+		XFREE(MTYPE_TMP, hash_data);
+		return NULL;
+	}
+
+	return hash_data;
+}
+
+static struct zebra_tc_class *tc_class_release(struct zebra_tc_class *qdisc,
+					       bool free_data)
+{
+	struct zebra_tc_class *lookup;
+
+	lookup = hash_lookup(zrouter.class_hash, qdisc);
+
+	if (!lookup)
+		return NULL;
+
+	return tc_class_free(lookup, free_data);
+}
+
+void zebra_tc_class_add(struct zebra_tc_class *class)
+{
+	if (IS_ZEBRA_DEBUG_TC)
+		zlog_debug("%s: add tc class ifindex %d kind %s", __func__,
+			   class->class.ifindex,
+			   tc_qdisc_kind2str(class->class.kind));
+
+	struct zebra_tc_class *found;
+	struct zebra_tc_class *old;
+	struct zebra_tc_class *new;
+
+	found = tc_class_lookup_handle(class);
```

Comment:
This is not correct indeed because handles are unique only per interface but not globally.

Reply:
In fact I think a maybe better way to do these lookups is: `hash_key = jhash(class.interface, class.handle)`, and `hash_equal` also only tells the difference of these two fields. Then by interface & handle we can locate the class immediately. We do extra `deep equal` to check equality and decide whether it is updated or newly added.